Critical Hits, Super effective, multi-hit moves dialogue

DustinLin opened the pull request:
### Fixes issue #680 
---- original behavior on `main`, on faint doesn't display effectiveness dialogue, or multi hit result:

https://github.com/pagefaultgames/pokerogue/assets/39450497/919527d3-7a3a-4ea4-8c84-7f17b746be67

---- PR fix: critical hits, effectiveness, and multi-hit results (even if ended early due to faint) how properly show when fainting

https://github.com/pagefaultgames/pokerogue/assets/39450497/f392e989-6380-4e34-9b4f-2d68631979eb


---- some more bugs to do?

For moves like absorb that have subsequent dialogue, it runs into the same issue of not rendering on faint, but it was not fixed with this PR, perhaps this can be done in a separate issue, needs more investigating. More generally the `end()` function of the `MoveEffectPhase` class in `phases.ts` doesn't queue message properly on faint, it seems like those messages are lose

https://github.com/pagefaultgames/pokerogue/assets/39450497/0d346fdd-dc1a-40a4-8a67-5fa3563340b4

molker commented on src/phases.ts:
```diff
@@ -3000,7 +3004,8 @@ export class PostTurnStatusEffectPhase extends PokemonPhase {
 export class MessagePhase extends Phase {
-  private text: string;
+  // todo change backlk to private
+  public text: string;
```
does this need to be public still?

molker commented on src/field/pokemon.ts:
```diff
@@ -1680,17 +1703,19 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
-      this.scene.unshiftPhase(new FaintPhase(this.scene, this.getBattlerIndex(), preventEndure));
-      this.resetSummonData();
-    }
+    // checks and adds Fainted scene
```
should this be moved up with the check

DustinLin commented on src/field/pokemon.ts:
```diff
@@ -1680,17 +1703,19 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
     const damagePhase = new DamagePhase(this.scene, this.getBattlerIndex(), damage, result as DamageResult, critical);
     this.scene.unshiftPhase(damagePhase);
+    // queue critical message before effectiveness
+    if (critical)
```
probably not, for readability reasons I should move it back up right before the effectiveness so they are right next to each other

DustinLin commented:
thanks for the help!

DustinLin commented:
Believe I have found the fixes to my "more Bugs to do", will update the PR shortly

DustinLin commented:
Found a cleaner solution to the multi-hit dialogue, as well as resolving the problems referenced above in `---- some more bugs to do?`

new current display:



https://github.com/pagefaultgames/pokerogue/assets/39450497/e039d054-395a-41bc-80ac-ea9efb7a7dce


https://github.com/pagefaultgames/pokerogue/assets/39450497/ac513dc7-121f-4d12-ac95-cc2002ea1723

josericardo-fo commented:
LGTM but it has some conflicts

DustinLin commented:
seems like the conflicts are easy to resolve: simply take the incoming change, notice that the if/switch statement has already been defined above by me


for future reference, who deals with merge conflicts, the reviewers or the PR author?

bennybroseph commented:
The author

DustinLin commented:
fixed!

bennybroseph commented on src/overrides.ts:
need to remove overrides from commits

DustinLin replied:
updated/took what is currently on `main`, seems like molker did some reordering

karl-police commented:
Now, I read up the other things.

I will write shortly, but I just noticed it never told me, that Earthquake has no effect on Flying Pokemon.

DustinLin commented:
thanks for such a extensive write up karl, the intention is to show the number if hits **before** the fainting. I'll look into reproducing your set-up

karl-police commented:
Now I tried, Leftovers and Multi-Lense

With the opponent having spore and rough skin and I gave myself leftovers.


Idk wtf I did
![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/7b3b7546-5577-4488-964c-0397fac25481)

Game froze here, could be a different bug
![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/cca39660-91d7-4d54-96bd-ef8a6966a64c)
![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/cca39660-91d7-4d54-96bd-ef8a6966a64c)

![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/7b3b7546-5577-4488-964c-0397fac25481)

karl-police commented:
> Now I tried, Leftovers and Multi-Lense
> 
> With the opponent having spore and rough skin and I gave myself leftovers.
> 
> Idk wtf I did ![image](https://private-user-images.githubusercontent.com/12023782/332141617-7b3b7546-5577-4488-964c-0397fac25481.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyMjQ2NzksIm5iZiI6MTcxNjIyNDM3OSwicGF0aCI6Ii8xMjAyMzc4Mi8zMzIxNDE2MTctN2IzYjc1NDYtNTU3Ny00NDg4LTk2NGMtMDM5N2ZhYzI1NDgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIwVDE2NTkzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyNWRiNWY3YjVkNWNhMDNlN2IyMTVjMDA4NTZkYjJkOTMxZDlkODYxMzE1MmYzYTdlOTkzYTAwODhjMzEyNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hSNmFQkKU3yJvAgE1IhkHqIgqhR_AJ8eqqK5WrKYi-w)
> 
> Game froze here, could be a different bug ![image](https://private-user-images.githubusercontent.com/12023782/332141695-cca39660-91d7-4d54-96bd-ef8a6966a64c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyMjQ2NzksIm5iZiI6MTcxNjIyNDM3OSwicGF0aCI6Ii8xMjAyMzc4Mi8zMzIxNDE2OTUtY2NhMzk2NjAtOTFkNy00ZDU0LTk2YmQtZWY4YTY5NjZhNjRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIwVDE2NTkzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhMmU1YTg0ZTRjNmMwMjk1MDMwYjZiOWRkZjRlMTIxNWI4Mzc5NDhhMmVkMGE2MmYzZjk2ZGYwYTEwZGY4YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CV7pXTLziQ2uAQRExlkYGWEjjzhc4Kwz_bz1Bs-kbMc)
> 
> See in the console
> 
> It ended the TurnPhase, then it started another one 🤷



> Now I tried, Leftovers and Multi-Lense
> 
> With the opponent having spore and rough skin and I gave myself leftovers.
> 
> Idk wtf I did ![image](https://private-user-images.githubusercontent.com/12023782/332141617-7b3b7546-5577-4488-964c-0397fac25481.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyMjQ2NzksIm5iZiI6MTcxNjIyNDM3OSwicGF0aCI6Ii8xMjAyMzc4Mi8zMzIxNDE2MTctN2IzYjc1NDYtNTU3Ny00NDg4LTk2NGMtMDM5N2ZhYzI1NDgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIwVDE2NTkzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyNWRiNWY3YjVkNWNhMDNlN2IyMTVjMDA4NTZkYjJkOTMxZDlkODYxMzE1MmYzYTdlOTkzYTAwODhjMzEyNzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hSNmFQkKU3yJvAgE1IhkHqIgqhR_AJ8eqqK5WrKYi-w)
> 
> Game froze here, could be a different bug ![image](https://private-user-images.githubusercontent.com/12023782/332141695-cca39660-91d7-4d54-96bd-ef8a6966a64c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyMjQ2NzksIm5iZiI6MTcxNjIyNDM3OSwicGF0aCI6Ii8xMjAyMzc4Mi8zMzIxNDE2OTUtY2NhMzk2NjAtOTFkNy00ZDU0LTk2YmQtZWY4YTY5NjZhNjRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIwVDE2NTkzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhMmU1YTg0ZTRjNmMwMjk1MDMwYjZiOWRkZjRlMTIxNWI4Mzc5NDhhMmVkMGE2MmYzZjk2ZGYwYTEwZGY4YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CV7pXTLziQ2uAQRExlkYGWEjjzhc4Kwz_bz1Bs-kbMc)
> 
> See in the console
> 
> It ended the TurnPhase, then it started another one 🤷

Oh I think I get it maybe.

@DustinLin 
So, the opponent Pokemon had Rough Skin. I am not sure if the game froze, because my Pokemon died, but the attack still wanted to run.

DustinLin commented:
that seems like a different bug, if a pokemon dies it gets it's `BattleScene` removed, if that same pokemon was queued to do something, then referencing it's `null` scene could lead to the behavior you just got

karl-police commented:
Spore never triggered, so I removed Rough Skin now, and now I will test that. Maybe Spore is not a passive, idk.

Actually I got rid of Spore and just made it ``Species.AMOONGUSS`` and put spore on passive and ability

started a new run

used fury swipes spore triggered

and yeah it continued, but I think this PR is not about fixing that, so it shouldn't matter

karl-police commented:
@DustinLin 

So the only thing that I am unsure about, is the Flying Pokemon in the Double Battle not showing that the attack had no effect 🤷


![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/bbc656bf-bec6-4d57-8b1c-5c731dfc377a)

karl-police commented:
Then I tried self-destruct
![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/4b7d4d0a-f794-496b-b88b-7cbc69057104)

The K.O message never showed up. 

And my game froze and I managed to make it to the next wave

![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/79a07151-faee-41d1-a182-9b55023aa2c0)

@DustinLin 

![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/311e9ebc-0088-45d0-a873-c8272a01bf6e)
![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/f0a171d6-421d-46bc-9900-86f2901bc684)

karl-police commented:
> thanks for such a extensive write up karl, the intention is to show the number if hits **before** the fainting. I'll look into reproducing your set-up. In the original PR post i've included it properly displays the info in that order, I wonder what happened...

I don't know. I am using Github Desktop. And clicked on the pull and then it did the rest.

DustinLin commented:
@karl-police i figured out the bug behind the no-effect behavior, it has to do with conditioning the messages on if the move actually does damage or not (~line 1645 in `pokemon.ts`), I should probably include the fix in this commit, but it seems like the very same code is refactored in PR #924 (and seems to work properly in that case) so I'm not sure what to do.

The self-destruct is definitely something that needs to be fixed so I'll look into that, thanks!

karl-police commented:
> @karl-police i figured out the bug behind the no-effect behavior, it has to do with conditioning the messages on if the move actually does damage or not (~line 1645 in `pokemon.ts`), I should probably include the fix in this commit, but it seems like the very same code is refactored in PR #924 (and seems to work properly in that case) so I'm not sure what to do.
> 
> The self-destruct is definitely something that needs to be fixed so I'll look into that, thanks!

I'd ask @bennybroseph 

If he replies.

francktrouillez commented:
Thanks a lot for all the work you're doing! This has been bugging me since I started playing (even though it doesn't prevent me from enjoying the game!).

While testing the changes proposed in the PR, I found a new bug, which is the pokemon doesn't faint anymore if he dies from a status effect. I believe we might have some other new ones, since `#damageAndUpdate` is used quite a lot in different places.
Video:
https://github.com/pagefaultgames/pokerogue/assets/57403591/036c6a2a-c2c5-4807-966d-e1ebe6620ce8

I'm proposing in this PR a way to resolve the order of messages only. I'm not taking care of the multi-hit moves (I believe it's still something we want, but we can maybe have it in separate PRs if that makes sense).
- https://github.com/pagefaultgames/pokerogue/pull/1182

DustinLin commented:
@karl-police I believe I have fixed the game breaking self-faint bug you pointed out. I've tested and things still seem to be in order, it would be great to get some double confirmation.

The multi-hit dialogue is a larger issue, I'm proposing that for this PR (so that we can get it in and subsequent things merged and settled) we just include these changes. It could be the case that subsequent PRs solve it, but I'm doubtful.

The issue is essentially that the multi-hit message phase is being added *too late*, and other combat resolving code that is being run is messing with the order in which Phases are being added to the PhaseQueue. It will take me a bit more time to figure out everything that is going on there before I start messing around with that code, so I want to leave that for another issue/PR.

bennybroseph commented:
Long thread going on, which is good for progress. Hard to process though. Could we get a recap of what bugs still exist in this PR if any and what is still being worked on?

karl-police commented:
> @karl-police I believe I have fixed the game breaking self-faint bug you pointed out. I've tested and things still seem to be in order, it would be great to get some double confirmation.
> 
> The multi-hit dialogue is a larger issue, I'm proposing that for this PR (so that we can get it in and subsequent things merged and settled) we just include these changes. It could be the case that subsequent PRs solve it, but I'm doubtful.
> 
> The issue is essentially that the multi-hit message phase is being added _too late_, and other combat resolving code that is being run is messing with the order in which Phases are being added to the PhaseQueue. (more details are in the comments of the code of the commit) I have a few ideas for how to fix it, but it will take me a bit more time to figure out everything that is going on there before I start messing around with that code, so I want to leave that for another issue/PR.

I think the message appearing too late is not as bad as not appearing at all.


Eitherways, idk what is wrong now
![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/903c1f9c-1866-4347-af33-2f09977b9d75)

karl-police commented:
Immunity message thingy works now

![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/9b425e3b-f8d9-47bd-b556-b40664b13f3a)

However, it says that it doesn't effect the Steel Pokemon over there. But that's only because it's floating. Shouldn't it have a different message?


Self-Destruct works too, but the K.O. messages don't appear? Not sure if it's ment to have one.

karl-police commented:
> I think that is the intended dialogue for levitate? It seems like the ability is showing, and i'm guessing the dialogue reads "it has no effect".
> 
> As to @bennybroseph 's message to recap: there were a couple problems karl-police found when testing
> 
> 1. The multi hit dialogue is still out of order - I've found the root cause of this issue but it may take a bit to fix. I think the best plan is to include it in another PR (as karl-police mentioned it's better than not appearing at all at the moment).
> 2. There seems to be a couple bugs regarding the items Multi-lens and recovery seed? (Paraphrasing karl-police here) This seems to be a separate issue and not related to the changes here though.

When I looked at Phaser and these "Phases", my first thought was _How do actual Pokemon games code it_. Because it all seems very very dank that everything is encapsulated in these Phases.

I think aslong there's no other bugs, it should be fine. Recovery Seed works, when I use Self-Destruct it doesn't revive me, that could be a different issue, maybe even intended.

DustinLin commented:
Thanks for reminding me, but I was the one who posted a writeup explaining `Phase`s and how they work in the discord a couple weeks ago, I can maybe add that documentation in another PR, or wherever @bennybroseph thinks would be a good fit for it

karl-police commented:
Where's that K.O. message though? I think that's the only remaining mystery.

DustinLin commented:
I think that may be for 1-hit KO moves? Like fissure. Explosion/self-destruct are not one-hit KO moves so i don't think they classify. I haven't played any of the newer main series games, but I don't think a KO dialogue appears?

DustinLin commented:
Hey all, I've merged in the latest changes from `main`, which includes PR #1182 from @francktrouillez . many of our changes are similar, so I fixed the conflicts and added/removed from both of ours. I believe I also fixed the multi-hit dialogue order, I've tested all of @karl-police 's cases and they seem to work out fine? If any of you 2 could also confirm this that would be great.

bennybroseph commented:
Have we also tested Multi-Lens?

DustinLin commented:
Here is a demo of a regular move tackle

https://github.com/pagefaultgames/pokerogue/assets/39450497/198bd03a-bf12-4d02-9367-b7bf3c8b25c5


I tested 3 special cases also:

1. Mutli-lens with absorb: Seems like intended dialogue, recall that effectiveness dialogue only displays on the last hit. The absorb dialogue displayed twice but I think that's intended/not a change from this PR

https://github.com/pagefaultgames/pokerogue/assets/39450497/4b4c8e10-879e-4ba3-8127-83be48604dc4

2. Multi-lens with explosion: doesn't explode twice and faints as intended

https://github.com/pagefaultgames/pokerogue/assets/39450497/85f061a9-4964-459e-86e7-f155fccfc876


4. Multi-lens with bullet seed: seems like the multi-hit count stacks as shown in the clip below, we get a total of 6 hits (took some RNG to get this)

https://github.com/pagefaultgames/pokerogue/assets/39450497/bc9764b6-9635-483b-9618-68690c3b4889

karl-police commented:
> Have we also tested Multi-Lens?

Yes, but I think changes were made.

karl-police commented:
Were changes made, to be re-tested?

karl-police commented:
Testing things.

![image](https://github.com/pagefaultgames/pokerogue/assets/12023782/4064defb-5034-44c8-9317-1388acaf15d0)

This issue here, is something different. It asks me to restart the game, even though the Pokemon's faint animation didn't play. But that's a like Self Destruct related thing or something.

DustinLin commented:
what is this case? I tested something similar with explosion and it seems to be ok. I've never got this "retry" message before

karl-police commented:
I am not sure how Earthquake in a double battle just attacked my own Pokemon that used it 🤷

karl-police commented:
> I tested double battle with Rough Skin and Poison Point and it works.

If I use Fury Swipe and the enemy Pokemon has Rough Skin. If I die, I won't see the hit count, while I have this prompt enabled that tells me if I want to retry the Battle.

Also, Earthquake attacked my own Pokemon that used it in a double battle, I am not exactly sure how that happened, but it happened once only.

karl-police commented:
Self-destruct issue was fixed though.

DustinLin commented:
I can reproduce your fury-swipe + rough skin combo, and yes it doesn't display if you faint yourself. IIRC this is what also happens in the main series games so it's ok?

For the earthquake thing (which may be a separate issue?) I can't tell the specific set up, is it something like this below, which seems to work?

https://github.com/pagefaultgames/pokerogue/assets/39450497/4c171425-e578-4b56-bad9-3cb9193b8bee

karl-police commented:
> I can reproduce your fury-swipe + rough skin combo, and yes it doesn't display if you faint yourself. IIRC this is what also happens in the main series games so it's ok?
> 
> For the earthquake thing (which may be a separate issue?) I can't tell the specific set up, is it something like this below, which seems to work?
> 
>  Screen.Recording.2024-05-23.at.12.09.00.PM.mov

sorta

Charmander needs to use the attack and faint. And then your other Pokemon needs to use Earthquake after.

DustinLin commented:
hmm I tried again and still can't reproduce it, maybe I'm doing something wrong still, but let's think of this in a separate issue/PR

https://github.com/pagefaultgames/pokerogue/assets/39450497/e91d380b-c0c1-4a1f-ac47-f51022990cd2

karl-police commented on src/field/pokemon.ts:
```diff
@@ -1858,18 +1868,39 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
+
+
+        /*
+        if (damage){
           this.scene.clearPhaseQueueSplice();
 
           const attacker = this.scene.getPokemonById(source.id);
           destinyTag?.lapse(attacker, BattlerTagLapseType.CUSTOM);
         }
```
Pokemon has so many attacks and features. I wonder if Phaser + Scenes is a good way to implement them without losing oversight or organized code.

DustinLin commented:
did some quick testing, original changes from the PR are still good to go

CodeTappert commented:
@DustinLin The github pages workflow is failing. Please fix your code. 
![image](https://github.com/pagefaultgames/pokerogue/assets/38758606/5b1142e5-0541-45c6-90d8-636733eaa440)

DustinLin commented:
thanks for the notif, fixed

DustinLin commented:
had some merge/branch issues and the `wind_rider` test failing for some reason, but the tests all are good now!
